File: src/main/java/com/github/kory33/signvote/ui/player/stats/TotalVoteStats.java
package com.github.kory33.signvote.ui.player.stats;

import java.util.Comparator;

import com.github.kory33.signvote.collection.RunnableInvoker;
import org.bukkit.entity.Player;

import com.github.kory33.signvote.configurable.JSONConfiguration;
import com.github.kory33.signvote.constants.StatsType;
import com.github.kory33.signvote.manager.PlayerInteractiveInterfaceManager;
import com.github.kory33.signvote.model.VotePointStats;
import com.github.kory33.signvote.session.VoteSession;
import com.github.kory33.signvote.ui.player.model.BrowseablePageInterface;

/**
 * Implementation of stats interface which sorts vote point statistics by total vote number.
 * @author Kory
 */
public final class TotalVoteStats extends StatsInterface {
    TotalVoteStats(Player player, VoteSession targetVoteSession, JSONConfiguration messageConfiguration,
                   RunnableInvoker runnableInvoker, PlayerInteractiveInterfaceManager interfaceManager, int pageIndex) {
        super(player, targetVoteSession, messageConfiguration, runnableInvoker, interfaceManager, pageIndex);
    }

    private TotalVoteStats(TotalVoteStats oldInterface, int newIndex) {
        super(oldInterface, newIndex);
    }


    @Override
    protected BrowseablePageInterface yieldPage(int pageIndex) {
        return new TotalVoteStats(this, pageIndex);
    }

    @Override
    public StatsType getStatsType() {
        return StatsType.VOTES;
    }

    @Override
    public Comparator<? super VotePointStats> getStatsComparator() {
        return Comparator.comparing(VotePointStats::getTotalVotes, Comparator.reverseOrder());
    }

    @Override
    public Number getStatsDisplayedValue(VotePointStats stats) {
        return stats.getTotalVotes();
    }
}

File: src/main/kotlin/com/github/kory33/signvote/ui/player/stats/TotalVoteStats.kt
package com.github.kory33.signvote.ui.player.stats

import com.github.kory33.chatgui.command.RunnableInvoker
import com.github.kory33.chatgui.manager.PlayerInteractiveInterfaceManager
import com.github.kory33.chatgui.model.player.IBrowseablePageInterface
import com.github.kory33.signvote.configurable.JSONConfiguration
import com.github.kory33.signvote.constants.StatsType
import com.github.kory33.signvote.session.VoteSession
import com.github.kory33.signvote.vote.VotePointStats
import org.bukkit.entity.Player

import java.util.Comparator
import java.util.function.Function

/**
 * Implementation of stats interface which sorts vote point statistics by total vote number.
 * @author Kory
 */
class TotalVoteStats : StatsInterface {
    internal constructor(player: Player, targetVoteSession: VoteSession, messageConfiguration: JSONConfiguration,
                         runnableInvoker: RunnableInvoker, interfaceManager: PlayerInteractiveInterfaceManager, pageIndex: Int) : super(player, targetVoteSession, messageConfiguration, runnableInvoker, interfaceManager, pageIndex)

    private constructor(oldInterface: TotalVoteStats, newIndex: Int) : super(oldInterface, newIndex)


    override fun yieldPage(pageIndex: Int): IBrowseablePageInterface {
        return TotalVoteStats(this, pageIndex)
    }

    override val statsType: StatsType
        get() = StatsType.VOTES

    override val statsComparator: Comparator<in VotePointStats>
        get() = Comparator.comparing(Function { it.totalVotes }, Comparator.reverseOrder<Int>())

    override fun getStatsDisplayedValue(stats: VotePointStats): Number {
        return stats.totalVotes
    }
}
